commitlint/grammar.md

```diff
@@ -0,0 +1,112 @@
+# Mender conventional commit specification
+
+```
+<type>(<scope>): <subject>
+<BLANK LINE>
+<body>
+<BLANK LINE>
+<footer>
+```
+
+# Specification
+
+    1. Commits MUST be prefixed with a type, which consists of a noun, feat, fix, etc., followed by the OPTIONAL scope, and REQUIRED terminal colon and space (See #allowed-types for more information).
+    2. The type feat MUST be used when a commit adds a new feature to your application or library.
+    3. The type fix MUST be used when a commit represents a bug fix for your application.
+    4. A scope MAY be provided after a type. A scope MUST consist of a noun describing a section of the codebase surrounded by parenthesis, e.g., fix(parser):
+    5. A description MUST immediately follow the colon and space after the type/scope prefix. The description is a short summary of the code changes, e.g., fix: array parsing issue when multiple spaces were contained in string.
+    6. A longer commit body MAY be provided after the short description, providing additional contextual information about the code changes. The body MUST begin one blank line after the description.
+    7. A commit body is free-form and MAY consist of any number of newline separated paragraphs.
+    8. One or more footers MAY be provided one blank line after the body. Each footer MUST consist of a word token, followed by a :<space> separator, followed by a string value (this is inspired by the git trailer convention). See #allowed-footers for more information.
+    9. A footer’s token MUST use - in place of whitespace characters, e.g., Acked-by (this helps differentiate the footer section from a multi-paragraph body). An exception is made for BREAKING CHANGE, which MAY also be used as a token.
+    10. A footer’s value MAY contain spaces and newlines, and parsing MUST terminate when the next valid footer token/separator pair is observed.
+    11. Breaking changes MUST be indicated in the footer.
+    12. A breaking change MUST consist of the uppercase text BREAKING CHANGE, followed by a colon, space, and description, e.g., BREAKING CHANGE: environment variables now take precedence over config files.
+    13. Types other than feat and fix MAY be used in your commit messages, e.g., docs: updated ref docs (See #allowed-types for more information).
+    14. BREAKING-CHANGE MUST be synonymous with BREAKING CHANGE, when used as a token in a footer.
+    15. A feat, fix or a BREAKING CHANGE must have a Ticket, and a Changelog in the footer. The allowed values for the Changelog is either freeform text, or `None`. The allowed value for `Ticket` is a valid ticket reference.
```

Comment:
Changelog yes, but I'm less sure about ticket... There are quite a lot of fixes that don't have ticket numbers.

I think I would lean towards making ticket optional, at least for "fix", but I'd argue for all three. For instance when we went from Mender client 1.7 to 2.0, this was primarily because of the renaming of the flag `-rootfs` to `-install`, and then again we did `-install` to `install` when we went from 2.x to 3.0. These were triggered by other work we were doing, but were not directly related to any ticket.

What do you think?

Reply:
I see and understand your argument - however I disagree.

I think that having it enforced makes sure that we never miss one, and enforcing it is a good reminder in all instances it is forgotten. It simply gives me peace of mind.

I understand that it is possibly adding a fair bit of typing, but this is something your editor can help you our with, and is what it is for, right? :smiley:

Reply:
> I think that having it enforced makes sure that we never miss one

Hmm, but are you thinking of `Ticket: none`? This could be a good idea.

However, in the text you are writing that a ticket number is *required*. Hence you'd have to go to Jira and create dummy tickets for every fix you make.

Reply:
Ohh... Good eye! 

Fixing